quic/s2n-quic-platform/src/message/cmsg.rs

```diff
@@ -4,173 +4,248 @@
+        unsafe {
+            debug_assert!(
+                align_of::<T>() <= align_of::<cmsghdr>(),
+                "alignment of T should be less than or equal to cmsghdr"
+            );
 
-        // IP_PKTINFO
-        #[cfg(s2n_quic_platform_pktinfo)]
-        {
-            len += libc::CMSG_LEN(
-                size_of::<libc::in_pktinfo>().max(size_of::<libc::in6_pktinfo>()) as _,
-            ) as usize;
-        }
+            // CMSG_SPACE() returns the number of bytes an ancillary element
+            // with payload of the passed data length occupies.
+            let element_len = size_of_cmsg::<T>();
+            debug_assert_ne!(element_len, 0);
+            debug_assert_eq!(libc::CMSG_SPACE(size_of::<T>() as _) as usize, element_len);
 
-        len
-    }
+            let new_cursor = self.cursor.checked_add(element_len).ok_or(OutOfSpace)?;
 
-    let len = unsafe { tx_len().max(rx_len()) };
+            self.storage
+                .len()
+                .checked_sub(new_cursor)
+                .ok_or(OutOfSpace)?;
 
-    // We use the MAX_LEN to determine if the cmsg has been populated at all so the actual
-    // len must be less than it, rather than less than or equal.
-    assert!(
-        len < MAX_LEN,
-        "required len should be less than maximum allocated len"
-    );
+            let cmsg_ptr = {
+                // Safety: the msg_control buffer should always be allocated to MAX_LEN
+                let msg_controllen = self.cursor;
+                let msg_control = self.storage.as_mut_ptr().add(msg_controllen as _);
+                msg_control as *mut cmsghdr
+            };
+
+            {
+                let cmsg = &mut *cmsg_ptr;
+
+                // interpret the start of cmsg as a cmsghdr
+                // Safety: the cmsg slice should already be zero-initialized and aligned
+
+                // Indicate the type of cmsg
+                cmsg.cmsg_level = level;
+                cmsg.cmsg_type = ty;
+
+                // CMSG_LEN() returns the value to store in the cmsg_len member
+                // of the cmsghdr structure, taking into account any necessary
+                // alignment.  It takes the data length as an argument.
+                cmsg.cmsg_len = libc::CMSG_LEN(size_of::<T>() as _) as _;
+            }
+
+            {
+                // Write the actual value in the data space of the cmsg
+                // Safety: we asserted we had enough space in the cmsg buffer above
+                // CMSG_DATA() returns a pointer to the data portion of a
+                // cmsghdr. The pointer returned cannot be assumed to be
+                // suitably aligned for accessing arbitrary payload data types.
+                // Applications should not cast it to a pointer type matching the
+                // payload, but should instead use memcpy(3) to copy data to or
+                // from a suitably declared object.
+                let data_ptr = cmsg_ptr.add(1);
```

Comment:
not sure I'm following why its adding 1 here?

Reply:
It's what the [`CMSG_DATA`](https://docs.rs/libc/latest/src/libc/unix/linux_like/mod.rs.html#1552) function does, except ours is actually sound since we're not going from a `*const` to a `*mut`. I've including the call to CMSG_DATA in a debug_assertion to make sure they return the same thing.